### DISCUSSION TITLE
How do you keep your app size under control ?

### DISCUSSION THREAD

**Discussion by vakhidbetrakhmadov** [MobileNativeFoundation/discussions#215] [General]
Hello everyone !

I would like to start a conversation on how different teams keep their app size under control, specifically on strategies, processes and tools employed to both prevent uncontrolled growth and to reduce already bloated app size.

I would in particular be interested in iOS, but Android experiences would be very interesting as well.

**Comment by vakhidbetrakhmadov** [comment]
For iOS in my team:
- We monitor app install size (as displayed by App Store). As part of our release pipeline, we pull app install size from App Store Connect and stores it in our BI, so we can see how app size changed with each release.
- We run static code analysis. As part of our automation, once a weak we run SwiftLint and custom scripts to detect unused code/resources, forgotten feature flags, etc and report them to the owners.
- We apply optimisations. We periodically shift our focus from other things to app size and explore new "magic" ways to reduce app size with comparatively little overall changes (striping symbols, tweaking build setting, etc)

But we currently have no preventative controls as you can see, so we are exploring ways to detect and block needless app size increments on pull requests.

I've read about how some teams use emerge tools for this purpose, for instance there is [this](https://engineering.atspotify.com/2023/11/the-what-why-and-how-of-mastering-app-size/) very interesting blogpost from Spotify team, where they explain how on each PR they build 2 version of the app, head and base, upload to emerge for analyses and then block PR based on provided feedback.

I am afraid introducing similar check for us might not be very affordable with our current CI capacity, because that's basically x2 app builds and we live in a monorepo with ~7M LOC and multiple apps, though we do use Bazel, and that might probably help us to reduce the impact.

Do you check app size increments on PR ? What strategies and tools do you employ ?

**Reply by BalestraPatrick** [reply]
FYI a recent improvement at Spotify was not to run twice the builds on every PR, but install reuse the cached result of an app size report for a master build. This allows to still do the comparison pretty efficiently, but at most compute the app size report once for every revision.

**Comment by rockbruno** [comment]
We did a talk at SwiftConf last year showing how we do this at Spotify: https://www.youtube.com/watch?v=v3rYaEXzRh4

I must say though that even with all those precautions it's quite hard to keep iOS app size in check because Swift itself is really inefficient in this regard.
